automates/program_analysis/PyAST2CAST/py_ast_to_cast.py

```diff
@@ -0,0 +1,1321 @@
+
+    def visit_ClassDef(self, node: ast.ClassDef):
+        """Visits a PyAST ClassDef node, which is used to define user classes.
+        Acquiring the fields of the class involves going through the __init__
+        function and seeing if the attributes are associated with the self
+        parameter. In addition, we add to the 'classes' dictionary the name of 
+        the class and a list of all its functions.
+
+        Args:
+            node (ast.ClassDef): A PyAST class definition node
+
+        Returns:
+            ClassDef: A CAST class definition node
+        """
+
+        name = node.name
+        self.classes[name] = []
+
+        bases = []
+        for base in node.bases:
+            bases.extend(self.visit(base))
+        
+        funcs = []
+        for func in node.body:
+            funcs.extend(self.visit(func))
+            if isinstance(func,ast.FunctionDef):
+                self.classes[name].append(func.name)
+        
+        fields = []
+
+        # Get the fields in the class
+        init_func = None
+        for f in node.body:
+            if isinstance(f,ast.FunctionDef) and f.name == "__init__":
+                init_func = f.body
+                break
+
+        for func_node in init_func:
+            if isinstance(func_node,ast.Assign) and isinstance(func_node.targets[0],ast.Attribute):
+                attr_node = func_node.targets[0]
+                if attr_node.value.id == "self":
+                    ref = [SourceRef(source_file_name=self.filenames[-1], col_start=attr_node.col_offset, col_end=attr_node.end_col_offset, row_start=attr_node.lineno, row_end=attr_node.end_lineno)]
+                    fields.append(Var(Name(attr_node.attr, source_refs=ref), "integer", source_refs=ref))
+
+        ref = [SourceRef(source_file_name=self.filenames[-1], col_start=node.col_offset, col_end=node.end_col_offset, row_start=node.lineno, row_end=node.end_lineno)]
+        return [ClassDef(name, bases, funcs, fields, source_refs=ref)]
+
+    def visit_Compare(self, node: ast.Compare):
+        """Visits a PyAST Compare node, which consists of boolean operations
+
+        Args:
+            node (ast.Compare): A PyAST Compare node
+
+        Returns:
+            BinaryOp: A BinaryOp node, which in this case will hold a boolean
+            operation
+        """
+
+        ops = {
+            ast.And: BinaryOperator.AND,
+            ast.Or: BinaryOperator.OR,
+            ast.Eq: BinaryOperator.EQ,
+            ast.NotEq: BinaryOperator.NOTEQ,
+            ast.Lt: BinaryOperator.LT,
+            ast.LtE: BinaryOperator.LTE,
+            ast.Gt: BinaryOperator.GT,
+            ast.GtE: BinaryOperator.GTE,
+            ast.In: BinaryOperator.IN,
+            ast.NotIn: BinaryOperator.NOTIN,
+        }
+
+        # Fetch the first element (which is in left)
+        left = node.left
+
+        # Grab the first comparison operation using reverse
+        node.ops.reverse()
```

Comment:
I was definitely not aware that you could pass indices (let alone negative ones) into pop()! Okay I'm restructuring this bit so that it doesn't use reverse then. My mistake.
Thank you for that!